engine/runtime/src/main/java/org/enso/interpreter/node/callable/argument/ReadArgumentCheckNode.java

```diff
@@ -0,0 +1,175 @@
+package org.enso.interpreter.node.callable.argument;
+
+import com.oracle.truffle.api.CompilerAsserts;
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.MaterializedFrame;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.ExplodeLoop;
+import com.oracle.truffle.api.nodes.InvalidAssumptionException;
+import com.oracle.truffle.api.nodes.Node;
+import java.util.Arrays;
+import java.util.stream.Collectors;
+import org.enso.interpreter.node.EnsoRootNode;
+import org.enso.interpreter.node.callable.ApplicationNode;
+import org.enso.interpreter.node.callable.InvokeCallableNode.DefaultsExecutionMode;
+import org.enso.interpreter.node.expression.builtin.meta.IsValueOfTypeNode;
+import org.enso.interpreter.node.expression.builtin.meta.TypeOfNode;
+import org.enso.interpreter.node.expression.literal.LiteralNode;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.callable.UnresolvedConversion;
+import org.enso.interpreter.runtime.callable.argument.CallArgument;
+import org.enso.interpreter.runtime.callable.function.Function;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.error.DataflowError;
+import org.enso.interpreter.runtime.error.PanicException;
+import org.graalvm.collections.Pair;
+
+public abstract class ReadArgumentCheckNode extends Node {
+  private final String name;
+  @Child IsValueOfTypeNode checkType;
+  @CompilerDirectives.CompilationFinal(dimensions = 1)
+  private final Type[] expectedTypes;
+  @CompilerDirectives.CompilationFinal
+  private String expectedTypeMessage;
+
+  ReadArgumentCheckNode(String name, Type[] expectedTypes) {
+    this.name = name;
+    this.checkType = IsValueOfTypeNode.build();
+    this.expectedTypes = expectedTypes;
+  }
+
+  static ReadArgumentCheckNode build(String argumentName, Type[] expectedTypes) {
+    if (expectedTypes == null || expectedTypes.length == 0) {
```

Comment:
That's a bit unusual. IIRC for all nodes we assume that `build` will return the node, no exceptions?